### PR TITLE
[TPU] XLA changes for finetuning

### PR DESCRIPTION
**Pull request by gkroiz**
This PR adds changes specific for finetuning on TPUs. 
Used TPU v4-8 with
```
log_interval = 1
devices = 4
batch_size = 64 / devices
micro_batch_size = 4
gradient_accumulation_steps = 4
num_epochs = 5
```

for stablelm-base-alpha-3b; resulting finetuning time ~= 4500s.

### CODE REVIEW

**Comment by gkroiz** [comment]
Would it be simpler to have a separate file for tpu finetuning (for example, `adapter_tpu.py`)?

**Comment by gkroiz** [comment]
cc @Liyang90 for review

**Comment by gkroiz** [comment]
TODO: if we decide to keep these changes in `adapter.py`, we also need to make the same adjustments in `adapter_v2.py`

File: finetune/adapter.py [hunk 1]
```diff
@@ -117,13 +140,22 @@ def train(
 
         input_ids, targets = get_batch(fabric, train_data)
 
-        with fabric.no_backward_sync(model, enabled=((iter_num + 1) % gradient_accumulation_steps != 0)):
+        if fabric.device.type == "xla":
             logits = model(input_ids)
             loss = loss_fn(logits, targets)
             fabric.backward(loss / gradient_accumulation_steps)
+        else:
+            with fabric.no_backward_sync(model, enabled=((iter_num + 1) % gradient_accumulation_steps != 0)):
```
Comment (carmocca) [inline finetune/adapter.py:148]: Why are we avoiding this? If there's an issue, we should fix it on Fabric.

File: finetune/adapter.py [hunk 2]
```diff
@@ -117,13 +140,22 @@ def train(
...
+                logits = model(input_ids)
+                loss = loss_fn(logits, targets)
+                fabric.backward(loss / gradient_accumulation_steps)
 
         if (iter_num + 1) % gradient_accumulation_steps == 0:
-            optimizer.step()
+            if fabric.device.type == "xla":
+                xm.optimizer_step(optimizer)
```
Comment (carmocca) [inline finetune/adapter.py:155]: Same question. Why can't we `optimizer.step()`?

File: finetune/adapter.py [hunk 3]
```diff
@@ -47,18 +48,34 @@
...
+    strategy = (DeepSpeedStrategy(config=ds_config) if (devices > 1 and not tpu) else "auto")
     fabric = L.Fabric(
-        devices=devices, strategy=(DeepSpeedStrategy(config=ds_config) if devices > 1 else "auto"), precision=precision
+        devices=devices, strategy=strategy, precision=precision
     )
-    fabric.launch()
+    
+    if tpu:
+        p_main = functools.partial(main, data_dir, checkpoint_dir, out_dir)
+        fabric._strategy._sync_module_states = False
```
Comment (carmocca) [inline finetune/adapter.py:65]: Why not instantiate with `XLAStrategy(sync_module_states=False)` if `tpu is True`?

**Review by carmocca — COMMENTED** [review]
Some early questions!

File: finetune/adapter.py (hunk 3 above)
Comment (gkroiz) [reply finetune/adapter.py:65]: This makes more sense, I will add a commit fixing this.

File: finetune/adapter.py (hunk 2 above)
Comment (gkroiz) [reply finetune/adapter.py:155]: [Here](https://github.com/pytorch/xla/blob/master/torch_xla/core/xla_model.py#L878-L911) is the src code for `xm.optimizer_step()`.

Using `xm.optimizer_step()` will also reduce the gradient. Otherwise we need to include `xm.reduce_gradients(optimizer)`

Comment (carmocca) [reply finetune/adapter.py:155]: In Fabric, `optimizer.step()` is wrapped by the `_FabricOptimizer`. This wrappers routes the call to `Strategy.optimizer_step` which itself routes it to `Precision.optimizer_step`.

Particularly, XLA on Fabric uses the `XLAPrecision` class which does exactly this:
https://github.com/Lightning-AI/lightning/blob/master/src/lightning/fabric/plugins/precision/xla.py

Comment (gkroiz) [reply finetune/adapter.py:157]: Hm, are we sure that the `xla` strategy wraps the optimizer? From what I can tell it is not wrapped by anything. I can create a PR in lightning src code if need be.

Comment (gkroiz) [reply finetune/adapter.py:157]: For reference when setting up everything, `XLAStrategy` and `ParallelStrategy` do not have any functions for setting up the optimizers so we default to `setup_optimizer()` in the base `Strategy` class: https://github.com/Lightning-AI/lightning/blob/master/src/lightning/fabric/strategies/strategy.py#L149.

Comment (carmocca) [reply finetune/adapter.py:157]: This happens through `fabric.setup`: https://github.com/Lightning-AI/lightning/blob/master/src/lightning/fabric/fabric.py#L223

If you are positive that we are not doing this wrapper correctly and calling `xm.optimizer_step()`, then there's a bug to fix upstream

Comment (gkroiz) [reply finetune/adapter.py:157]: Okay, the issue was that the precision for xla must be set to "32-true", but by default lit-parrot defines precision as "bf16-true". With "32-true" `xm.optimizer_step()` will be called as expected. I will make the excepted change.

**Comment by gkroiz** [comment]
When using TPUs, the current code in `adapter.py` won't work when using more than 8 cores (4 chips). 

For example, let's say we want to run finetune on `v4-64` (64 cores, 32 chips). We would need two device counts, one for local devices within each of the 8 workers and one for the total device count. The reasoning here is that the `XLA` fabric strategy needs to define `devices=4` (local device count) but for `batch_size`, `max_iters`, and `warmup_iters`, we would want to define based on the total device count, which would be 32. For my local tests, I made this distinction by defining `local_devices` and `total_devices`, but I'm not sure if this leads to unnecessary confusion when using non-xla strategies. What are your thoughts @carmocca?

File: finetune/adapter.py [hunk 4]
```diff
@@ -153,11 +153,8 @@ def train(
                 fabric.backward(loss / gradient_accumulation_steps)
 
         if (iter_num + 1) % gradient_accumulation_steps == 0:
-            if fabric.device.type == "xla":
-                xm.optimizer_step(optimizer)
-                xm.mark_step()
-            else:
-                optimizer.step()
+            optimizer.step()
+            if fabric.device.type == "xla": xm.mark_step()
```
[the remark this answers is not in the document]
Comment (gkroiz) [reply finetune/adapter.py:157]: Fabric calls this behind the scenes: https://github.com/Lightning-AI/lightning/blob/master/src/lightning/fabric/plugins/precision/xla.py

[the remark this answers is not in the document]
Comment (Liyang90) [reply finetune/adapter.py:157]: Oh I see. The `optimizer` is a Lightning wrapped one.

File: finetune/adapter.py [hunk 5]
```diff
@@ -47,18 +48,36 @@
...
+        devices=devices, strategy=strategy, precision="32-true" if tpu else precision
     )
-    fabric.launch()
+    
+    if tpu:
+        p_main = functools.partial(main, data_dir, checkpoint_dir, out_dir)
+        fabric.launch(p_main)
+    else:
+        fabric.launch()
+        main(data_dir, checkpoint_dir, out_dir, fabric)
```
Comment (carmocca) [inline finetune/adapter.py:71]: We can use the launched function pattern for both cases

```suggestion
    fabric.launch(main, data_dir, checkpoint_dir, out_dir)
```

File: finetune/adapter.py [hunk 6]
```diff
@@ -195,7 +228,7 @@ def pad_right(x, pad_id):
     x = torch.stack([pad_right(x, pad_id=0) for x in input_ids])
     y = torch.stack([pad_right(x, pad_id=-1) for x in labels])
 
-    if isinstance(fabric.accelerator, MPSAccelerator):
+    if isinstance(fabric.accelerator, MPSAccelerator) or fabric.device.type == 'xla':
```
Comment (carmocca) [inline finetune/adapter.py:231]: We can simplify here:

```suggestion
    if fabric.device.type in ("mps", "xla"):
```

**Review by carmocca — COMMENTED** [review]
I think making these distinctions is fine. Because they are still true in the general multi-node case, not just with XLA.

**Comment by gkroiz** [comment]
@carmocca I noticed this PR is now closed. Do we no longer need to xla changes?

**Comment by carmocca** [comment]
Sorry, it was an accident!

File: finetune/adapter.py [hunk 7]
```diff
@@ -48,18 +47,31 @@
...
-        devices=devices, strategy=(DeepSpeedStrategy(config=ds_config) if devices > 1 else "auto"), precision=precision
+    strategy = (
+        "auto"
+        if devices <= 1
+        else XLAStrategy(sync_module_states=False) if tpu else DeepSpeedStrategy(config=ds_config)
     )
-    fabric.launch()
+    # For multi-host TPU training, the device count for Fabric is limited to the count on a single host.
+    fabric_devices = "auto" if (tpu and devices > 1) else devices
+    fabric = L.Fabric(devices=fabric_devices, strategy=strategy, precision="32-true" if tpu else precision)
```
Comment (carmocca) [inline finetune/adapter.py:64]: 16-mixed would be a valid argument too for XLA.

I would suggest that we do instead

```
precision: Optional[str] = None


if precision is None:
    precision = "32-true" if tpu else "16-true"
```

So that Fabric can raise an error for any other value instead of silently choosing 32-true

**Review by carmocca — APPROVED** [review]
Minor comment. Rest looks good

File: finetune/adapter.py (hunk 7 above)
Comment (gkroiz) [reply finetune/adapter.py]: Thanks for the suggestion, just pushed this change.

**Comment by gkroiz** [comment]
Looks like this PR did not include changes from https://github.com/Lightning-AI/lit-parrot/pull/127. Will need to create a small follow-up PR fixing the adapter files.
